More utils

fgrunewald opened the pull request.

pckroon commented on cgsmiles/graph_utils.py:
```diff
@@ -180,3 +180,47 @@ def set_atom_names_atomistic(molecule, meta_graph=None):
+    netwokrx.Graph
+    """
+    meta_graph = nx.Graph()
+    fragments = defaultdict(list)
```
```suggestion
    fragments = defaultdict(set)
```
?

fgrunewald replied:
Removed `fragments` entirely — it was write-only (populated but never read anywhere in the function), so the list-vs-set question was moot. Fixed in fe12a5e.

pckroon commented on cgsmiles/graph_utils.py:
```diff
@@ -180,3 +180,47 @@ def set_atom_names_atomistic(molecule, meta_graph=None):
+    for node in molecule.nodes:
+        unique_values = molecule.nodes[node][unique_attr]
+        for unique_val in unique_values:
+            if unique_val in fragments:
+                fragments[unique_val].append(node)
+            else:
+                fragments[unique_val].append(node)
+                new_attrs = {attr: molecule.nodes[node][attr] for attr in copy_attrs}
+                new_attrs[unique_attr] = unique_val
+                meta_graph.add_node(node_counter, **new_attrs)
+                node_to_unique_value[unique_val] = node_counter
+                node_counter += 1
```
Do you want/need to check that all the copied over attributes are the same for the resulting meta nodes?
I don't love the variable name `node_counter`. Maybe `meta_node_idx`?

fgrunewald replied:
Renamed to `meta_node_idx`, and added an assertion that a repeat atom belonging to an already-created fragment agrees with the first atom's copied attributes (e.g. `fragname`) instead of silently ignoring a mismatch. Fixed in fe12a5e.

fgrunewald commented:
@pckroon this is very much in the draft stages at the momen. BUT it should fix the issue with ambiguity. If you make a cgsmiles string that is correct, resolve it and then from the aa molecule extract the fragments again, which is now possible, it should return a compressed but valid SMILES string. Just the get_fragment_dict function is currently an unholy abomination that needs some trimming. 

Tests fail because of the writer being incompatible with current pysmiles.

pckroon commented on cgsmiles/graph_utils.py:
```diff
@@ -218,34 +220,41 @@ def make_meta_graph(molecule, unique_attr='fragid', copy_attrs=['fragname']):
+        for u1, u2 in itertools.combinations(unique_values, r=2):
+            n1 = node_to_unique_value[u1]
+            n2 = node_to_unique_value[u2]
+            if meta_graph.has_edge(n1, n2):
+                meta_graph.edges[(n1, n2)]['order'] += 1
+            else:
+                meta_graph.add_edge(n1, n2, order=1)
```
Abstract this to an `add_edge` function. You also use it later (`if u1 != u2`)

fgrunewald replied:
Factored out `_add_or_increment_edge()` and reused it in the squash loop and the real-bond loop (the `itertools.product` branch below keeps its different add-only-if-missing semantics, so left as is). Fixed in fe12a5e.

fgrunewald commented:
## Additional review pass on this branch

Beyond the three threads above, I ran a broader review of the full diff (`master...more_utils`) and found some more issues, three of them with a real, demonstrated impact. I've already fixed all of these locally (working tree only, not pushed yet) and re-run the full test suite (182 passed) plus the two repros below — flagging here for visibility/discussion before I push.

### Correctness bugs

**1. `_are_isomorphic` almost never lets isomorphic squash-based fragments condense** (`cgsmiles/extractor.py:112`)
```python
for bond in bonds:
    if any(["!" in b for b in bond]) and len(bond) > 1:
        return False
```
`bond` here is a single bonding-descriptor *string* (e.g. `"!01"`), so `len(bond) > 1` is true for basically every squash descriptor — it should be `len(bonds) > 1`, checking the *node's* descriptor count (the comment says the intent is to guard against "symmetric fragments with two asymmetric squash operators"). I demonstrated the impact directly: resolving `{[#TC5]1[#TC5][#TC5]1}.{#TC5=[!]ccc[!]}` and round-tripping it through `MoleculeFragmentExtractor` currently produces **three** separate fragment names (`TC5`, `TC5A`, `TC5B`) instead of condensing to one `TC5` — i.e. the extractor's stated purpose ("most condensed representation") silently fails for any ring/squash-based molecule. Fixed by checking `len(bonds)` on the whole per-node list instead of a single descriptor string; benzene now correctly condenses to one `TC5`.

**2. `annotate_fragments` misattributes atoms when a virtual node precedes a real fragment** (`cgsmiles/graph_utils.py:135`, root cause in `cgsmiles/resolve.py:409`)
`annotate_fragments` looks up `meta_graph`'s own `'fragid'` node attribute to index into the atom-level fragid groups. But `resolve()` stamps `meta_graph` fragid via a flat `enumerate(self.meta_graph.nodes)` (counting virtual nodes), while the atom-level fragid assigned in `resolve_disconnected_molecule`/`merge_graphs` implicitly *skips* virtual nodes (its counter only advances when a real fragment is actually merged in). Reproduced directly: resolving `{[#TC4].[#SP4][#SP4]}.{#SP4=OC[$]C[$]O}` (virtual `TC4` first) and calling `annotate_fragments(cg, aa)` on the resolver's own `(cg, aa)` pair attached the first real `SP4`'s 9 atoms to the virtual `TC4` meta node, the second `SP4`'s atoms to the wrong meta node, and left the true second `SP4` meta node with an **empty** atom set. Fixed by having `resolve()` number real fragments with the same skip-virtual counter `merge_graphs` uses, and giving virtual nodes distinct negative sentinels that can never collide with a real fragid.

**3. `_match_bonds` bond-order check is a no-op tautology** (`cgsmiles/extractor.py:18`)
```python
if j not in used_indices and item1[0] == item2[0] and item1[-1] == item1[-1]:
```
`item1[-1] == item1[-1]` compares `item1` to itself — always `True` — so only the direction character is actually checked, not the bond order. Fixed the obvious copy-paste typo (`item2[-1]`). Currently mostly masked by bug #1 (condensation rarely triggers at all), but becomes load-bearing once #1 is fixed and more fragments actually reach this code path. I also added an `assert matched_bonds is not None` at the one call site (it's documented to return `None` on failure but that was never checked) — with the typo fixed, this is provably unreachable given the precondition established by the caller's isomorphism check, so the assert documents that invariant rather than silently `TypeError`-ing if it's ever violated by a future change.

**4. Unbounded suffix generator instead of a fixed 27-character pool** (`cgsmiles/extractor.py:78,172,193`)
`condense_fragments` disambiguated same-named non-isomorphic fragments by popping from `"ZYXWVUTSRQPONMLKJIHGFEDCBA "` — a hardcoded 27-slot pool (including a magic trailing space for the "no suffix" case) that raises a bare `IndexError` once a fragname collects more than 27 distinct fragments. Replaced with a small generator yielding the standard spreadsheet-column sequence (`A, B, ..., Z, AA, AB, ...`) with no cap.

### Cleanup

**5. Dead code in `annotate_bonding_operators`** (`cgsmiles/graph_utils.py:314-344`) — `sorted_nodes`/`index_map`/`sorted_edges` were computed with a comment claiming they're "required for the compression", but the loop right below iterates `nx.edge_dfs(molecule)`, never `sorted_edges` — plus a half-finished, fully commented-out `toggle`/`tdx` alternation mechanism. Removed all of it; behavior is unchanged (it was already dead), this just stops the comment promising an ordering guarantee the code doesn't provide.

**6. Redundant state in `make_meta_graph`** (`cgsmiles/graph_utils.py:226`) — `ref_values` (a list) duplicated exactly what `node_to_unique_value` (a dict, populated one line later in the same branch) already tracks as keys. Dropped it in favor of an O(1) `in node_to_unique_value` check instead of an O(n) list scan.

### Also flagged but *not* changed (lower confidence / discussion-worthy, not fixed)

- `cgsmiles/graph_utils.py:269` — the `itertools.product` branch of `make_meta_graph`'s final edge loop (squash-adjacent real bonds) adds a meta-edge only if missing, never incrementing `order`, unlike the otherwise-parallel squash loop and first branch (which do increment via `_add_or_increment_edge`). Plausible under a fused-ring layout combining a squash junction with a second real connector between the same two beads, but I didn't have a concrete repro and it may be intentional (avoiding inflated order from `product()` revisiting the same pair) — flagging for @fgrunewald to weigh in rather than guessing.
- `cgsmiles/extractor.py:169` — `condense_fragments` runs a full VF2 subgraph-isomorphism search against every previously-accepted same-named fragment with no cheap pre-filter (node/edge count, or the already-computed `nhash`), which could get expensive for long homopolymers. Performance suggestion, not a correctness issue — leaving as a follow-up rather than touching the matching logic in this pass.

All of this is implemented and verified locally (182/182 existing tests, plus the two repros above). Let me know if you'd like me to push it as-is, split it into separate commits, or hold off on any of it.
